File: Leetcode/linkedlistcycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


# value가 유일하다는 조건이 없으므로 head 자체(메모리를 저장해야됨)
# O(N)
class Solution:
    def hasCycle(self, head) -> bool:
        node_dict = {}
        while head != None:
            if node_dict.get(head):
                return True

            node_dict[head] = 1
            head = head.next
        return False

File: Leetcode/linkedlistcycle.cpp
// space complexity O(1)

#include <iostream>

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

using namespace std;

bool hasCycle(ListNode *head)
{
    ListNode *walker = head;
    ListNode *runner = head;

    while (runner && runner->next)
    {
        walker = walker->next;
        runner = runner->next->next;
        // runner가 먼저 끝나서 null 되면 사이클 없음
        // 사이클이 있다면 계속 돌 것
        if (runner == walker)
            return true;
    }
    return false;
}